tests/sparseml/core/lifecycle/test_event.py

```diff
@@ -0,0 +1,565 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from functools import partial
+
+import pytest
+
+from sparseml.core.event import Event, EventType
+from sparseml.core.lifecycle.event import (
+    CallbacksEventLifecycle,
+    EventLifecycle,
+    WrappedOptimEventLifecycle,
+)
+
+
+def test_event_lifecycle_can_not_be_instantiated():
+    with pytest.raises(TypeError):
+        EventLifecycle(type_first=EventType.BATCH_START, start=Event())
+
+
+class EventLifecycleDummyChild(EventLifecycle):
+    def batch_start_events(self):
+        return [], "batch_start_events"
+
+    def loss_calculated_events(self):
+        return [], "loss_calculated_events"
+
+    def optim_pre_step_events(self):
+        return [], "optim_pre_step_events"
+
+    def optim_post_step_events(self):
+        return [], "optim_post_step_events"
+
+    def batch_end_events(self):
+        return [], "batch_end_events"
+
+
+def _get_event_lifecycle(
+    start=None, type_first=None, lifecycle_class=EventLifecycleDummyChild
+):
+    start = start or Event()
+    type_first = type_first or EventType.BATCH_START
+    lifecycle = lifecycle_class(type_first=type_first, start=start)
+    return lifecycle
+
+
+class TestEventLifecycle:
+    def test_init(self):
+        start = Event()
+        type_first = EventType.BATCH_START
+
+        lifecycle = _get_event_lifecycle(type_first=type_first, start=start)
+        assert lifecycle.type_first == type_first
+        assert lifecycle.steps_per_epoch == start.steps_per_epoch
+        assert lifecycle.batches_per_step == start.batches_per_step
+        assert lifecycle.invocations_per_step == start.invocations_per_step
+        assert lifecycle.global_step == start.global_step
+        assert lifecycle.global_batch == start.global_batch
```

Comment:
Could we initialize these attributes in the start Event to test non-defaults?

Reply:
Yes, good call out! parameterized `type_first`, and `start`